Softeq.XToolkit.PushNotifications/PushNotificationsServiceBase.cs

```diff
@@ -9,91 +9,83 @@ namespace Softeq.XToolkit.PushNotifications
+        public async Task<PushNotificationsUnregisterResult> UnregisterFromPushNotifications(bool unregisterInSystem = false)
         {
-            if (string.IsNullOrWhiteSpace(token))
-            {
-                OnFailedToRegisterForPushNotifications("OnRegistered with empty token");
-            }
-            else
+            // Unregister in system if needed
+            if (PushTokenStorageService.IsTokenRegisteredInSystem && unregisterInSystem)
             {
-                OnRegisterSuccessInternal(token).SafeTaskWrapper(Logger);
+                var unregistered = await UnregisterFromPushTokenInSystem();
+                PushTokenStorageService.IsTokenRegisteredInSystem = !unregistered;
             }
-        }
-
-        public void OnFailedToRegisterForPushNotifications(string errorMessage)
-        {
-            Logger.Warn($"Push Notifications failed to register: {errorMessage}");
-            OnRegisterFailedInternal().SafeTaskWrapper(Logger);
-        }
-
-        private async Task OnRegisterSuccessInternal(string token)
-        {
-            PushTokenStorageService.IsTokenRegisteredInSystem = true;
 
-            if (PushTokenStorageService.PushToken != token || !PushTokenStorageService.IsTokenSavedOnServer)
+            var token = PushTokenStorageService.PushToken;
+            if (string.IsNullOrEmpty(token))
             {
-                PushTokenStorageService.PushToken = token;
-                await RegisterPushTokenOnServer();
+                return PushNotificationsUnregisterResult.Success;
             }
 
-            PushNotificationsHandler.OnPushRegistrationCompleted(PushTokenStorageService.IsTokenRegisteredInSystem, PushTokenStorageService.IsTokenSavedOnServer);
-        }
+            // Try to unregister on server
+            var removedFromServerStatus = await RemotePushNotificationsService
+                .RemovePushNotificationsToken(token).ConfigureAwait(false);
 
-        private async Task OnRegisterFailedInternal()
-        {
-            PushTokenStorageService.IsTokenRegisteredInSystem = false;
+            if (removedFromServerStatus == PushNotificationsUnregisterResult.Success)
+            {
+                // if we failed to remove it is left as it was (saved or not)
+                PushTokenStorageService.IsTokenSavedOnServer = false;
+            }
 
-            await UnregisterFromPushNotifications();
+            // Clear token if it is unregistered both in system and on server
+            if (!PushTokenStorageService.IsTokenRegisteredInSystem &&
+                !PushTokenStorageService.IsTokenSavedOnServer)
+            {
+                PushTokenStorageService.PushToken = string.Empty;
+            }
 
-            PushNotificationsHandler.OnPushRegistrationCompleted(PushTokenStorageService.IsTokenRegisteredInSystem, PushTokenStorageService.IsTokenSavedOnServer);
+            return removedFromServerStatus;
```

Comment:
Can this result be Failed? does the client have to define it in RemoteService?

Reply:
Fixed